Add FreeBSD container support

dfr opened the pull request:
Following on from #7472, this adds support for creating containers inside a pod sandbox. This gives us a functioning cri-o for FreeBSD, missing pod and container stats and some features such as device support and privileged containers. Part of #6492.

Note: the FreeBSD implementation of createSandboxContainer was copied from the Linux implementation and modified for FreeBSD. Some code could be shared between the two platforms at the expense of reduced readability. Its not clear to me which approach is better and I would welcome feedback on this.

<!--  Thanks for sending a pull request!

Please be aware that we're following the Kubernetes guidelines of contributing
to this project. This means that we have to use this mandatory template for all
of our pull requests.

Please also make sure you've read and understood our contributing guidelines
(https://github.com/cri-o/cri-o/blob/main/CONTRIBUTING.md) as well as ensuring
that all your commits are signed with `git commit -s`.

Here are some additional tips for you:

- If this is your first time, please read our contributor guidelines:
  https://git.k8s.io/community/contributors/guide#your-first-contribution and
  developer guide
  https://git.k8s.io/community/contributors/devel/development.md#development-guide
- Please label this pull request according to what type of issue you are
  addressing, especially if this is a release targeted pull request. For
  reference on required PR/issue labels, read here:
  https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
- If you want *faster* PR reviews, read how:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
- If the PR is unfinished, see how to mark it:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:

None

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes see:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Experimental support for FreeBSD.
```

haircommander commented:
I think I'd prefer to share code where possible. Even if the readibility suffers, it will reduce the liklihood we miss patches to one or the other

dfr commented:
I re-implemented this to keep as much of the code for container creation shared between Linux and FreeBSD. I kept the history fairly detailed to make it easier to see the sequence of small refactors clear.

haircommander commented on server/container_create_linux.go:
```diff
@@ -26,3 +27,25 @@ func (s *Server) createContainerPlatform(ctx context.Context, container *oci.Con
 	return s.Runtime().CreateContainer(ctx, container, cgroupParent, false)
 }
+
+func newLinuxContainerSecurityContext() *types.LinuxContainerSecurityContext {
```
nit can the function be defined below where it's called?

dfr commented:
Note: this iteration disables timezone support for FreeBSD - this can be removed if we update the vendor of github.com/containers/common to v0.58 but perhaps that should happen in a separate PR?

kwilczynski commented:
@dfr, you might have to rebase this against `main` as there have been changes that will affect your work here.

dfr commented:
> @dfr, you might have to rebase this against `main` as there have been changes that will affect your work here.

I just rebased and updated the PR, thanks!

haircommander commented:
there do seem to be some legitimate breakages [here](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/cri-o_cri-o/7727/pull-ci-cri-o-cri-o-main-ci-fedora-integration/1781000256628985856/artifacts/fedora-integration/cri-o-gather/artifacts/testout.txt)

sohankunkerkar commented:
@dfr, could you rebase on top of the main branch to pick the `container/common` changes?

dfr commented:
> @dfr, could you rebase on top of the main branch to pick the `container/common` changes?

Thanks for updating the vendor of c/common - I rebased this PR and it builds on FreeBSD with the timezone support enabled. I haven't tested that timezone actually works yet - I will take some time to re-test it tomorrow.

sohankunkerkar commented:
@dfr it looks like the following test is failing:
```sh
not ok 135 ctr log linking
# (in test file ./ctr.bats, line 1274)
#   `[ -f "$mounted_log_path" ]' failed
# time="2024-05-07T15:56:11Z" level=warning msg="Support for the 'registries' option has been dropped but it is referenced in \"/etc/crio/crio.conf.d/01-registries.conf\".  Please use containers-registries.conf(5) for configuring unqualified-search registries instead."
# time="2024-05-07 15:56:11.078674775Z" level=info msg="Starting CRI-O, version: 1.31.0, git: 87b94db74c4971f3a66eea890e77b2e768ad4712(dirty)"
# level=info msg="Using default capabilities: CAP_CHOWN, CAP_DAC_OVERRIDE, CAP_FSETID, CAP_FOWNER, CAP_SETGID, CAP_SETUID, CAP_SETPCAP, CAP_NET_BIND_SERVICE, CAP_KILL"
# time="2024-05-07 15:56:11.124499041Z" level=info msg="Starting CRI-O, version: 1.31.0, git: 87b94db74c4971f3a66eea890e77b2e768ad4712(dirty)"
# level=info msg="Using default capabilities: CAP_CHOWN, CAP_DAC_OVERRIDE, CAP_FSETID, CAP_FOWNER, CAP_SETGID, CAP_SETUID, CAP_SETPCAP, CAP_NET_BIND_SERVICE, CAP_KILL"
# time="2024-05-07T15:56:11Z" level=fatal msg="validate service connection: validate CRI v1 runtime API for endpoint \"unix:///tmp/tmp.BAWjzOkNiN/crio.sock\": rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /tmp/tmp.BAWjzOkNiN/crio.sock: connect: no such file or directory\""
# # --- crio.log :: ---
```

dfr commented:
/retest-required

sohankunkerkar commented:
@dfr could you rebase this PR on top of the main branch? It looks like CI is green now.
@cri-o/cri-o-maintainers PTAL

dfr commented:
> @dfr could you rebase this PR on top of the main branch? It looks like CI is green now. @cri-o/cri-o-maintainers PTAL

I just rebased and pushed so hopefully the CI stays green. I build it locally for both linux and freebsd and did some very light testing on freebsd.

dfr commented:
/retest-required

dfr commented:
/retest-required

dfr commented:
So failing tests are mostly ci/prow (which never seem to pass for me). The outlier is codecov/patch is showing reduced coverage which is confusing - is it possible that moving things around has caused that or is it just the new FreeBSD code which is not tested by the Linux test suite.

I did some more extensive testing with a FreeBSD crio built from this PR and it works well enough to run a single node FreeBSD Kubernetes cluster.

dfr commented:
/retest-required

dfr commented:
/retest

haircommander commented:
/retest

I think this is fixed now, though if not you may need (another) rebase. thanks for your seeming unyielding patience!

dfr commented:
/retest

sohankunkerkar commented:
/retest

dfr commented:
/retest

dfr commented:
This is down to two failing tests, ci/prow/ci-fedora-critest and ci/prow/ci-fedora-integration which seem unrelated. I will try to re-run them and see if they pass today.

dfr commented:
Looking at this again this morning, all the tests are passing. Is anything blocking this from being merged? This is the largest part of my FreeBSD changes and once its merged, subsequent pull requests should much smaller.

sohankunkerkar reviewed (approved):
/lgtm

sohankunkerkar commented:
@cri-o/cri-o-maintainers PTAL

haircommander commented on server/container_create_common.go:
```diff
@@ -0,0 +1,969 @@
+	}
+}
+
+func (s *Server) createSandboxContainer(ctx context.Context, ctr ctrfactory.Container, sb *sandbox.Sandbox) (cntr *oci.Container, retErr error) {
```
maybe we move stuff in this file to `server/container_create.go`?

haircommander commented:
sorry one more structural nit but otherwise LGTM, I would suggest we do it as a follow-up but it will make the git history harder to work with

dfr commented:
> sorry one more structural nit but otherwise LGTM, I would suggest we do it as a follow-up but it will make the git history harder to work with

Just to make sure I understand correctly, you are suggesting moving common code from container_create_linux.go to container_create.go instead of creating container_create_common.go? That makes sense - I will rework the PR along those lines tomorrow.

haircommander commented:
yup exactly! and while you're at it, would you make sure that functions you're moving are above any functions they call? I like to be able to read the file from top to bottom, descending down the stack

dfr commented:
> yup exactly! and while you're at it, would you make sure that functions you're moving are above any functions they call? I like to be able to read the file from top to bottom, descending down the stack

I re-worked the PR to move shared logic to container_create.go as discussed.

dfr commented:
/retest

dfr commented:
/retest

sohankunkerkar commented:
/retest

dfr commented:
/retest

dfr commented:
/retest

dfr commented:
/retest-required

dfr commented:
The integration/integration tests seem to be timeout related? I was concerned about the integration/critest failures but I have been able to run critest locally on a ubuntu VM with crun so I'm not sure what's going on there.

dfr commented:
/retest-required

dfr commented:
As far as I can tell, all required tests are passing. What can I do to help help this pull request to be merged? If there is too much changing in a single pull request, I could split it up into a set of smaller pull requests although the last in that sequence would still be large due to the code movement from container_create_linux.go to container_create.go to perhaps not that helpful.

Any ideas on alternative approaches that may be easier to review would be very welcome.

haircommander commented:
let's get this merged after we cut 1.31.0

dfr commented:
/retest

haircommander commented:
/retest

kwilczynski commented:
/hold

So I can have a review.

saschagrunert commented:
@dfr do we want to include this into 1.32 or can we move it into 1.33?

dfr commented:
> @dfr do we want to include this into 1.32 or can we move it into 1.33?

It would be nice to get this into 1.32 if possible - what are the cutoff dates for 1.32 or 1.33? I will aim to rebase this tomorrow.

On the other hand, while I do have one group which would like to test out FreeBSD-native kubernetes, they can always build cri-o from my github fork.

haircommander commented:
cut-off would be Dec 17th  :slightly_smiling_face:

dfr commented:
Rebased - this was a little time consuming due to conflicts with #8397 but that refactor ended up being quilte helpful and simplified this change in a couple of places.

Looks like I missed the cut for 1.32 - perhaps it would be good to merge early in the 1.33 cycle since this is still a fairly large change.

saschagrunert commented:
> Looks like I missed the cut for 1.32 - perhaps it would be good to merge early in the 1.33 cycle since this is still a fairly large change.

Not yet, but I agree it would be good to move this into 1.33.

dfr commented:
I will rebase it again this week and hopefully we can get this merged sometime soon.

dfr commented:
Rebased.

sohankunkerkar commented:
@dfr Thanks for your patience.
Could you fix this test?
https://github.com/cri-o/cri-o/actions/runs/13288331349/job/37102473662?pr=7727

dfr commented:
> @dfr Thanks for your patience. Could you fix this test? https://github.com/cri-o/cri-o/actions/runs/13288331349/job/37102473662?pr=7727

Done.

bitoku commented:
It's not clear to me what should be working with this PR.
How about adding an integration test for freebsd?

There's no default runner, but there seem some 3rd party options.
https://github.com/actions/runner/issues/385

I think it's ok to disable some tests, but we can get better understanding what is remaining to fully support freebsd.

dfr commented:
> It's not clear to me what should be working with this PR. How about adding an integration test for freebsd?
> 
> There's no default runner, but there seem some 3rd party options. [actions/runner#385](https://github.com/actions/runner/issues/385)
> 
> I think it's ok to disable some tests, but we can get better understanding what is remaining to fully support freebsd.

I think that at this stage, I would be happy with a simple 'does it build' test. We did this in Podman and it has been quite helpful - this was easy in Podman since they use Cirrus CI which supports FreeBSD. I'm not sure what it would take to add a 3rd party FreeBSD runner.

bitoku commented:
hmm, did you do any manual test on this change?
I suppose CreateContainer should succeed probably.

> I think that at this stage, I would be happy with a simple 'does it build' test.
I'm not sure what it would take to add a 3rd party FreeBSD runner.

I believe we should have the e2e test for the freebsd functionality as soon as possible, but you made the point about 3rd party runner. we may as well not implement it for now.

dfr commented:
> hmm, did you do any manual test on this change? or is this PR just a preliminary one for the future changes? I suppose CreateContainer should succeed probably.

I have a small set of tests which use crictl to create pods and containers in a few configurations that cover creating simple containers, containers with volumes, etc. and I run them manually. I also have a two node FreeBSD Kubernetes cluster that uses cri-o with this PR and a few much smaller patches that I can use for ad-hoc testing.

> 
> > I think that at this stage, I would be happy with a simple 'does it build' test.
> > I'm not sure what it would take to add a 3rd party FreeBSD runner.
> 
> I believe we should have the e2e test for the freebsd functionality as soon as possible, but you made the point about 3rd party runner. we may as well not implement it for now.

If we can collectively figure out a way to run tests in a FreeBSD VM, I would be very happy. Possibly https://github.com/marketplace/actions/cross-platform-action would work for this?

bitoku commented:
I created the issue for FreeBSD CI. https://github.com/cri-o/cri-o/issues/9005
Aside from that, as you tested it already manually, I feel ok to merge it.
Once you verified the binary, please let us know.

dfr commented:
> /lgtm /hold
> 
> Feel free to remove the hold when it's ready. Thanks for being so patient and thorough here. I'm very excited about this!

/unhold

After some more testing, including with my experimental FreeBSD Kubernetes cluster, I'm happy with this change. I will follow up with a few much smaller pull requests to fill out the features needed for minimal Kubernetes support.

dfr commented:
/retest-required

dfr commented:
/retest-required

dfr commented:
/retest-required

dfr commented:
> seems like unrelated timeouts and not related to these changes?

I think so - it usually takes a while to get this set of tests to pass. Unfortunately, they are important end-to-end tests ensuring that we can still bring up Kubernetes clusters etc. so we need them. I do wish they were less flaky though.

dfr commented:
/retest

sohankunkerkar commented:
/retest
